### ISSUE TITLE
[ipcamera] File descriptor leak

### ISSUE THREAD

**Issue by toffee** [openhab/openhab-addons#11301]
## Expected Behavior
After stopping the ipcamera binding the number of file descriptor used by openhab process should be the same as it was before starting the binding. 

## Current Behavior
If open the openhab fresh (without enabling any binding) the number of file descriptor is:
```bash
jupiter:~ # lsof -p 21250 | wc -l
392
jupiter:~ # lsof -p 21250 | grep FIFO | wc -l
94 
```

If re-open (that's why the PID is different) the openhab with ipcamera binding enabled the number of file descriptor is (greater than 1024):
```bash
jupiter:~ # lsof -p 24246 | wc -l
1364
jupiter:~ # lsof -p 24246 | grep FIFO | wc -l
737
```

After stopping the binding using the console ` bundle:stop org.openhab.binding.ipcamera` the number of file descriptor is (slight decrease but the number of FIFO remains the same):
```bash
jupiter:~ # lsof -p 24246 | wc -l
1351
jupiter:~ # lsof -p 24246 | grep FIFO | wc -l
737
```

I think there are two issues:
1. ipcamera opens a lot of FIFO file descriptor (above I show the total file descriptor number for openhab but also the number of FIFO). My configuration has 6 cameras, one using ffmpeg.
2. ipcamera does not cleanup the opened file descriptors on exit   

## Context
I'm using some bindings (modbus, digiplex) that benefits form ```serial``` binding services. Unfortunately the serial binding is affected by nrjavaserial (see NeuronRobotics/nrjavaserial#111) locking. Even if I recompiled the nrjavaserial with ```#define DISABLE_LOCKFILES``` the issue appear if ipcamera binding is started (and make openhab to have more that 1024 file descriptors open) before binding(s) using serial interface. 

My current understanding is that serial binding works only if openhab process has less that 1024 file descriptors open (or binding(s) using serial interface open the serial port when the number of file descriptor is less than [1024](https://github.com/NeuronRobotics/nrjavaserial/blob/7aa21d1dc8cecdc8daad3ebc40273cfb8179e9d2/src/main/c/src/SerialImp.c#L4204)), So, to be able to use ipcamera with other serial bindings it's necessary to keep ipcamera open file descriptors under control.

The ipcamera things in my configuration (4 outdoor Hikvision camera, 1 indoor Foscam camera and 1 door video intercom):
```
Thing ipcamera:hikvision:cam01 "Camera 01" @ "Cameras"
[
    ipAddress="192.168.120.51", 
    username="viewer",
    password="****",
    serverPort=10051
]

Thing ipcamera:hikvision:cam02 "Camera 02" @ "Cameras"
[
    ipAddress="192.168.120.52", 
    username="viewer",
    password="****",
    serverPort=10052
]

Thing ipcamera:hikvision:cam03 "Camera 03" @ "Cameras"
[
    ipAddress="192.168.120.53", 
    username="viewer",
    password="****",
    serverPort=10053
]

Thing ipcamera:hikvision:cam04 "Camera 04" @ "Cameras"
[
    ipAddress="192.168.120.54", 
    username="viewer",
    password="****",
    serverPort=10054
]

// Need to enalble the sub stream to be in MJPEG format
// http://192.168.120.57:88/cgi-bin/CGIProxy.fcgi?cmd=setSubStreamFormat&format=1&usr=xxx&pwd=xxx
Thing ipcamera:foscam:cam07 "Camera 07" @ "Cameras"
[
    ipAddress="192.168.120.57",
    port=88,
    username="viewer",
    password="****",
    serverPort=10057
]

Thing ipcamera:generic:cam09 "Camera 09" @ "Cameras"
[
    ipAddress="192.168.120.59",
    username="admin",
    password="****",
    serverPort=10059,
    snapshotUrl="ffmpeg",
    mjpegUrl="ffmpeg",
    ffmpegInput="rtsp://192.168.120.59/ch1/main/av_stream",
    ffmpegLocation="/usr/bin/ffmpeg"
]
```

## Environment
* Version used: version 3.1.0 for openHAB and ipcamera binding
* Operating System and version: openHAB running on docker 20.10.6-ce on openSUSE Leap 15.3

**Comment by Skinah** [comment]
I'll have to look into this in more depth over the coming days. Have you seen this thread?
https://community.openhab.org/t/zwave-and-ipcamera-issue/122209/42

I also did a quick search on Netty and it seems they did have a leak reported that was fixed so it may be a regression in the version that is being used currently. Thanks for reporting this.

**Comment by Skinah** [comment]
I ran the Karaf console command `shell:info` and it gives me the following with at least 14 bindings and 4 cameras online and idling:

 Open file descriptors       774
 Max file descriptors        102,642

Note that this is after I have made a lot of changes to the binding since the version you will be running. It also contains a breaking change where the URLS are different that you will need to be aware of, but since that has already been merged you can make the change now to it and not worry about it later.

Precompiled jar and also the source code if you wish to build it:
https://github.com/openhab/openhab-addons/pull/11300

Updated documentation that is updated to cover the URL changes:
https://next.openhab.org/addons/bindings/ipcamera/#video-streams

If you could re-test with these recent changes that would be great to see if it resolves the issue for you. I am running Netty 4.1.63.Final which is the normal version for 3.2m2

**Comment by Skinah** [comment]
OK I have done some testing and I had the issue with 996 open file descriptors (potentially GC lowered it down after the serial port broke). I made some changes as Netty preassigns a threadpool based on your (NUM CPU CORES * 2) which was a lot higher than needed. Others are reporting it is fixed for them with 3.2M3, but most of the changes that will help are not yet merged, see above PR.

To reproduce I put extra camera things in which I did not have and I could break the serial port by doing this when I rebooted. After these changes I could have 8 cameras without any issues.

I have 4 cameras and a group thing running at 450 OFD

**Comment by Skinah** [comment]
Thanks for confirming there is a big improvement on your system which confirms what I have found. Regarding the mjpeg streams not working on your HIK cameras, this should not be happening and it works here on my HIK cameras. One possible explanation is that some web browsers have a limit of how many streams they can display at the same time from the same IP and port. Now the binding is serving all cameras on the same port this limit can be reached if you are trying to display many cameras on the page at the same time. What happens if you try the url in a browser without any widgets running that may use cameras? A look at a single camera running in TRACE mode should help fault find. Just pause the other cameras first so the log is clean and only showing 1 camera.
